File: src/lib/tauri/menu.ts
import { listen } from "@tauri-apps/api/event";
import {
  Menu,
  type MenuItemOptions,
  type PredefinedMenuItemOptions,
  type SubmenuOptions
} from "@tauri-apps/api/menu";
import { invoke } from "@tauri-apps/api/core";
import { getCurrentWindow } from "@tauri-apps/api/window";
import { t, type AppLanguage, type I18nKey } from "../i18n";

export type NativeMenuHandlers = Partial<Record<NativeMenuCommand, () => unknown | Promise<unknown>>>;

export type NativeMenuCommand =
  | "openDocument"
  | "saveDocument"
  | "saveDocumentAs"
  | "formatBold"
  | "formatItalic"
  | "formatStrikethrough"
  | "formatInlineCode"
  | "formatParagraph"
  | "formatHeading1"
  | "formatHeading2"
  | "formatHeading3"
  | "formatBulletList"
  | "formatOrderedList"
  | "formatQuote"
  | "formatCodeBlock"
  | "insertLink"
  | "insertImage";

type NativeMenuCommandPayload = {
  command: NativeMenuCommand;
};

function runNativeMenuAction(handler: (() => unknown | Promise<unknown>) | undefined) {
  if (!handler) return;

  Promise.resolve(handler()).catch(() => {});
}

function invokeNativeWindowCommand(command: "open_blank_editor_window" | "open_settings_window") {
  invoke(command).catch(() => {});
}

function customItem(
  id: string,
  text: string,
  accelerator: string | undefined,
  handler: (() => unknown | Promise<unknown>) | undefined
): MenuItemOptions {
  return {
    id,
    text,
    accelerator,
    action: () => runNativeMenuAction(handler)
  };
}

function commandItem(
  id: string,
  text: string,
  accelerator: string | undefined,
  command: "open_blank_editor_window" | "open_settings_window"
): MenuItemOptions {
  return {
    id,
    text,
    accelerator,
    action: () => invokeNativeWindowCommand(command)
  };
}

function separator(): PredefinedMenuItemOptions {
  return {
    item: "Separator"
  };
}

function predefined(item: PredefinedMenuItemOptions["item"], text?: string): PredefinedMenuItemOptions {
  return text ? { item, text } : { item };
}

function menuLabel(language: AppLanguage, key: I18nKey) {
  return t(language, key);
}

async function isCurrentNativeWindowFocused() {
  try {
    return await getCurrentWindow().isFocused();
  } catch {
    return true;
  }
}

export async function listenNativeApplicationMenuCommands(handlers: NativeMenuHandlers) {
  return listen<NativeMenuCommandPayload>("markra://menu-command", async (event) => {
    if (!(await isCurrentNativeWindowFocused())) return;

    runNativeMenuAction(handlers[event.payload.command]);
  });
}

export function createNativeApplicationMenuItems(
  handlers: NativeMenuHandlers,
  language: AppLanguage = "en"
): SubmenuOptions[] {
  const label = (key: I18nKey) => menuLabel(language, key);
  const newDocument = commandItem("newDocument", label("menu.newDocument"), "CmdOrCtrl+N", "open_blank_editor_window");
  const openDocument = customItem("openDocument", label("menu.openDocument"), "CmdOrCtrl+O", handlers.openDocument);
  const saveDocument = customItem("saveDocument", label("menu.saveDocument"), "CmdOrCtrl+S", handlers.saveDocument);
  const saveDocumentAs = customItem(
    "saveDocumentAs",
    label("menu.saveDocumentAs"),
    "CmdOrCtrl+Shift+S",
    handlers.saveDocumentAs
  );
  const settings = commandItem("openSettings", label("menu.settings"), "CmdOrCtrl+,", "open_settings_window");
  const bold = customItem("formatBold", label("menu.bold"), "CmdOrCtrl+B", handlers.formatBold);
  const italic = customItem("formatItalic", label("menu.italic"), "CmdOrCtrl+I", handlers.formatItalic);
  const strikethrough = customItem(
    "formatStrikethrough",
    label("menu.strikethrough"),
    "CmdOrCtrl+Shift+X",
    handlers.formatStrikethrough
  );
  const inlineCode = customItem("formatInlineCode", label("menu.inlineCode"), "CmdOrCtrl+E", handlers.formatInlineCode);
  const paragraph = customItem("formatParagraph", label("menu.paragraph"), "CmdOrCtrl+Alt+0", handlers.formatParagraph);
  const heading1 = customItem("formatHeading1", label("menu.heading1"), "CmdOrCtrl+Alt+1", handlers.formatHeading1);
  const heading2 = customItem("formatHeading2", label("menu.heading2"), "CmdOrCtrl+Alt+2", handlers.formatHeading2);
  const heading3 = customItem("formatHeading3", label("menu.heading3"), "CmdOrCtrl+Alt+3", handlers.formatHeading3);
  const bulletList = customItem(
    "formatBulletList",
    label("menu.bulletList"),
    "CmdOrCtrl+Shift+8",
    handlers.formatBulletList
  );
  const orderedList = customItem(
    "formatOrderedList",
    label("menu.orderedList"),
    "CmdOrCtrl+Shift+7",
    handlers.formatOrderedList
  );
  const quote = customItem("formatQuote", label("menu.quote"), "CmdOrCtrl+Shift+B", handlers.formatQuote);
  const codeBlock = customItem("formatCodeBlock", label("menu.codeBlock"), "CmdOrCtrl+Alt+C", handlers.formatCodeBlock);
  const link = customItem("insertLink", label("menu.link"), "CmdOrCtrl+K", handlers.insertLink);
  const image = customItem("insertImage", label("menu.image"), "CmdOrCtrl+Shift+I", handlers.insertImage);

  return [
    {
      id: "markra:app",
      text: "Markra",
      items: [
        predefined({ About: { name: "Markra" } }),
        separator(),
        settings,
        separator(),
        predefined("Hide", label("menu.hide")),
        predefined("HideOthers", label("menu.hideOthers")),
        predefined("ShowAll", label("menu.showAll")),
        separator(),
        predefined("Quit", label("menu.quit"))
      ]
    },
    {
      id: "markra:file",
      text: label("menu.file"),
      items: [
        newDocument,
        openDocument,
        separator(),
        saveDocument,
        saveDocumentAs,
        separator(),
        predefined("CloseWindow", label("menu.closeWindow"))
      ]
    },
    {
      id: "markra:edit",
      text: label("menu.edit"),
      items: [
        predefined("Undo", label("menu.undo")),
        predefined("Redo", label("menu.redo")),
        separator(),
        predefined("Cut", label("menu.cut")),
        predefined("Copy", label("menu.copy")),
        predefined("Paste", label("menu.paste")),
        predefined("SelectAll", label("menu.selectAll"))
      ]
    },
    {
      id: "markra:format",
      text: label("menu.format"),
      items: [
        bold,
        italic,
        strikethrough,
        inlineCode,
        separator(),
        paragraph,
        heading1,
        heading2,
        heading3,
        separator(),
        bulletList,
        orderedList,
        quote,
        codeBlock,
        separator(),
        link,
        image
      ]
    },
    {
      id: "markra:view",
      text: label("menu.view"),
      items: [predefined("Fullscreen", label("menu.fullscreen"))]
    }
  ];
}

export async function installNativeApplicationMenu(handlers: NativeMenuHandlers, language: AppLanguage = "en") {
  const stopListening = await listenNativeApplicationMenuCommands(handlers);

  try {
    const menu = await Menu.new({
      items: createNativeApplicationMenuItems(handlers, language)
    });
    await menu.setAsAppMenu();
  } catch {
    // Keep the Rust-installed startup menu working if the JS menu API is unavailable.
  }

  return stopListening;
}

export function createNativeEditorContextMenuItems(handlers: NativeMenuHandlers, language: AppLanguage = "en") {
  const label = (key: I18nKey) => menuLabel(language, key);

  return [
    predefined("Cut", label("menu.cut")),
    predefined("Copy", label("menu.copy")),
    predefined("Paste", label("menu.paste")),
    predefined("SelectAll", label("menu.selectAll")),
    separator(),
    customItem("markra:context:bold", label("menu.bold"), "CmdOrCtrl+B", handlers.formatBold),
    customItem("markra:context:italic", label("menu.italic"), "CmdOrCtrl+I", handlers.formatItalic),
    customItem("markra:context:link", label("menu.link"), "CmdOrCtrl+K", handlers.insertLink)
  ];
}

export async function installNativeEditorContextMenu(
  target: Pick<EventTarget, "addEventListener" | "removeEventListener">,
  handlers: NativeMenuHandlers,
  language: AppLanguage = "en"
) {
  try {
    const menu = await Menu.new({
      items: createNativeEditorContextMenuItems(handlers, language)
    });

    const handleContextMenu = (event: Event) => {
      const element = event.target instanceof Element ? event.target : null;
      if (!element?.closest(".markdown-paper")) return;

      event.preventDefault();
      menu.popup();
    };

    target.addEventListener("contextmenu", handleContextMenu);

    return () => {
      target.removeEventListener("contextmenu", handleContextMenu);
    };
  } catch {
    return () => {};
  }
}

File: apps/desktop/src-tauri/src/menu.rs
use crate::language::{resolve_startup_language, AppLanguage};
use std::collections::HashMap;
use tauri::{
    menu::{AboutMetadata, Menu, MenuBuilder, MenuItemBuilder, SubmenuBuilder},
    Manager,
};

pub(crate) const NATIVE_MENU_COMMAND_EVENT: &str = "markra://menu-command";

const NEW_DOCUMENT_COMMAND: &str = "newDocument";
const SETTINGS_WINDOW_COMMAND: &str = "openSettings";

#[derive(Clone, serde::Serialize)]
pub(crate) struct NativeMenuCommand {
    pub(crate) command: String,
}

fn app_menu_item<R: tauri::Runtime, M: Manager<R>>(
    manager: &M,
    id: &str,
    text: &str,
    accelerator: &str,
) -> tauri::Result<tauri::menu::MenuItem<R>> {
    MenuItemBuilder::with_id(id, text)
        .accelerator(accelerator)
        .build(manager)
}

pub(crate) fn create_application_menu<R: tauri::Runtime>(
    app: &tauri::AppHandle<R>,
) -> tauri::Result<Menu<R>> {
    let language = resolve_startup_language(&app.config().identifier);
    create_application_menu_for_language(app, language, None)
}

fn create_application_menu_for_language<R: tauri::Runtime>(
    app: &tauri::AppHandle<R>,
    language: AppLanguage,
    accelerators: Option<&HashMap<String, String>>,
) -> tauri::Result<Menu<R>> {
    let labels = crate::menu_labels::for_language(language);

    let new = app_menu_item(
        app,
        NEW_DOCUMENT_COMMAND,
        labels.new_document,
        "CmdOrCtrl+N",
    )?;
    let open = app_menu_item(app, "openDocument", labels.open_document, "CmdOrCtrl+O")?;
    let close = app_menu_item(app, "closeDocument", labels.close_document, "CmdOrCtrl+W")?;
    let save = app_menu_item(app, "saveDocument", labels.save_document, "CmdOrCtrl+S")?;
    let save_as = app_menu_item(
        app,
        "saveDocumentAs",
        labels.save_document_as,
        "CmdOrCtrl+Shift+S",
    )?;
    let export_pdf = app_menu_item(app, "exportPdf", labels.export_pdf, "CmdOrCtrl+P")?;
    let export_html = app_menu_item(app, "exportHtml", labels.export_html, "CmdOrCtrl+Shift+E")?;
    let export_menu = SubmenuBuilder::with_id(app, "markra:file:export", labels.export)
        .items(&[&export_pdf, &export_html])
        .build()?;
    let settings = app_menu_item(app, SETTINGS_WINDOW_COMMAND, labels.settings, "CmdOrCtrl+,")?;

    let bold = app_menu_item(
        app,
        "formatBold",
        labels.bold,
        &menu_accelerator(accelerators, "formatBold", "CmdOrCtrl+B"),
    )?;
    let italic = app_menu_item(
        app,
        "formatItalic",
        labels.italic,
        &menu_accelerator(accelerators, "formatItalic", "CmdOrCtrl+I"),
    )?;
    let strikethrough = app_menu_item(
        app,
        "formatStrikethrough",
        labels.strikethrough,
        &menu_accelerator(accelerators, "formatStrikethrough", "CmdOrCtrl+Shift+X"),
    )?;
    let inline_code = app_menu_item(
        app,
        "formatInlineCode",
        labels.inline_code,
        &menu_accelerator(accelerators, "formatInlineCode", "CmdOrCtrl+E"),
    )?;
    let paragraph = app_menu_item(
        app,
        "formatParagraph",
        labels.paragraph,
        &menu_accelerator(accelerators, "formatParagraph", "CmdOrCtrl+Alt+0"),
    )?;
    let heading_1 = app_menu_item(
        app,
        "formatHeading1",
        labels.heading_1,
        &menu_accelerator(accelerators, "formatHeading1", "CmdOrCtrl+Alt+1"),
    )?;
    let heading_2 = app_menu_item(
        app,
        "formatHeading2",
        labels.heading_2,
        &menu_accelerator(accelerators, "formatHeading2", "CmdOrCtrl+Alt+2"),
    )?;
    let heading_3 = app_menu_item(
        app,
        "formatHeading3",
        labels.heading_3,
        &menu_accelerator(accelerators, "formatHeading3", "CmdOrCtrl+Alt+3"),
    )?;
    let bullet_list = app_menu_item(
        app,
        "formatBulletList",
        labels.bullet_list,
        &menu_accelerator(accelerators, "formatBulletList", "CmdOrCtrl+Shift+8"),
    )?;
    let ordered_list = app_menu_item(
        app,
        "formatOrderedList",
        labels.ordered_list,
        &menu_accelerator(accelerators, "formatOrderedList", "CmdOrCtrl+Shift+7"),
    )?;
    let quote = app_menu_item(
        app,
        "formatQuote",
        labels.quote,
        &menu_accelerator(accelerators, "formatQuote", "CmdOrCtrl+Shift+B"),
    )?;
    let code_block = app_menu_item(
        app,
        "formatCodeBlock",
        labels.code_block,
        &menu_accelerator(accelerators, "formatCodeBlock", "CmdOrCtrl+Alt+C"),
    )?;
    let link = app_menu_item(
        app,
        "insertLink",
        labels.link,
        &menu_accelerator(accelerators, "insertLink", "CmdOrCtrl+K"),
    )?;
    let image = app_menu_item(
        app,
        "insertImage",
        labels.image,
        &menu_accelerator(accelerators, "insertImage", "CmdOrCtrl+Shift+I"),
    )?;
    let table = app_menu_item(
        app,
        "insertTable",
        labels.table,
        &menu_accelerator(accelerators, "insertTable", "CmdOrCtrl+Alt+T"),
    )?;
    let toggle_file_list = app_menu_item(
        app,
        "toggleMarkdownFiles",
        labels.toggle_file_list,
        &menu_accelerator(accelerators, "toggleMarkdownFiles", "CmdOrCtrl+Shift+M"),
    )?;
    let toggle_markra_ai = app_menu_item(
        app,
        "toggleAiAgent",
        labels.toggle_markra_ai,
        &menu_accelerator(accelerators, "toggleAiAgent", "CmdOrCtrl+Alt+J"),
    )?;
    let ai_writing_command = app_menu_item(
        app,
        "toggleAiCommand",
        labels.ai_writing_command,
        &menu_accelerator(accelerators, "toggleAiCommand", "CmdOrCtrl+Shift+J"),
    )?;
    let toggle_source_mode = app_menu_item(
        app,
        "toggleSourceMode",
        labels.toggle_source_mode,
        &menu_accelerator(accelerators, "toggleSourceMode", "CmdOrCtrl+Alt+S"),
    )?;

    let app_menu = SubmenuBuilder::with_id(app, "markra:app", "Markra")
        .about(Some(AboutMetadata {
            name: Some("Markra".into()),
            ..Default::default()
        }))
        .separator()
        .items(&[&settings])
        .separator()
        .hide_with_text(labels.hide)
        .hide_others_with_text(labels.hide_others)
        .show_all_with_text(labels.show_all)
        .separator()
        .quit_with_text(labels.quit)
        .build()?;

    let file_menu = SubmenuBuilder::with_id(app, "markra:file", labels.file)
        .items(&[&new, &open, &close])
        .separator()
        .items(&[&save, &save_as])
        .separator()
        .items(&[&export_menu])
        .build()?;

    let edit_menu = SubmenuBuilder::with_id(app, "markra:edit", labels.edit)
        .undo_with_text(labels.undo)
        .redo_with_text(labels.redo)
        .separator()
        .cut_with_text(labels.cut)
        .copy_with_text(labels.copy)
        .paste_with_text(labels.paste)
        .select_all_with_text(labels.select_all)
        .build()?;

    let format_menu = SubmenuBuilder::with_id(app, "markra:format", labels.format)
        .items(&[&bold, &italic, &strikethrough, &inline_code])
        .separator()
        .items(&[&paragraph, &heading_1, &heading_2, &heading_3])
        .separator()
        .items(&[&bullet_list, &ordered_list, &quote, &code_block])
        .separator()
        .items(&[&link, &image, &table])
        .build()?;

    let view_menu = SubmenuBuilder::with_id(app, "markra:view", labels.view)
        .fullscreen_with_text(labels.fullscreen)
        .separator()
        .items(&[
            &toggle_file_list,
            &toggle_markra_ai,
            &ai_writing_command,
            &toggle_source_mode,
        ])
        .build()?;

    MenuBuilder::new(app)
        .items(&[&app_menu, &file_menu, &edit_menu, &format_menu, &view_menu])
        .build()
}

fn menu_accelerator(
    accelerators: Option<&HashMap<String, String>>,
    command: &str,
    fallback: &str,
) -> String {
    accelerators
        .and_then(|items| items.get(command))
        .filter(|accelerator| !accelerator.trim().is_empty())
        .cloned()
        .unwrap_or_else(|| fallback.to_string())
}

#[tauri::command]
pub(crate) fn install_application_menu(
    app: tauri::AppHandle,
    language: String,
    accelerators: Option<HashMap<String, String>>,
) -> Result<(), String> {
    let language = AppLanguage::from_code(&language)
        .ok_or_else(|| format!("Unsupported application menu language: {language}"))?;
    let menu = create_application_menu_for_language(&app, language, accelerators.as_ref())
        .map_err(|error| error.to_string())?;

    app.set_menu(menu)
        .map(|_| ())
        .map_err(|error| error.to_string())
}

pub(crate) fn is_native_new_window_command(command: &str) -> bool {
    command == NEW_DOCUMENT_COMMAND
}

pub(crate) fn is_native_settings_window_command(command: &str) -> bool {
    command == SETTINGS_WINDOW_COMMAND
}

pub(crate) fn is_frontend_menu_command(command: &str) -> bool {
    matches!(
        command,
        "openDocument"
            | "closeDocument"
            | "saveDocument"
            | "saveDocumentAs"
            | "exportPdf"
            | "exportHtml"
            | "formatBold"
            | "formatItalic"
            | "formatStrikethrough"
            | "formatInlineCode"
            | "formatParagraph"
            | "formatHeading1"
            | "formatHeading2"
            | "formatHeading3"
            | "formatBulletList"
            | "formatOrderedList"
            | "formatQuote"
            | "formatCodeBlock"
            | "insertLink"
            | "insertImage"
            | "insertTable"
            | "toggleMarkdownFiles"
            | "toggleAiAgent"
            | "toggleAiCommand"
            | "toggleSourceMode"
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::language::AppLanguage;

    #[test]
    fn menu_labels_follow_the_selected_startup_language() {
        let english = crate::menu_labels::for_language(AppLanguage::En);
        let simplified_chinese = crate::menu_labels::for_language(AppLanguage::ZhCn);

        assert_ne!(simplified_chinese.file, english.file);
        assert_ne!(simplified_chinese.new_document, english.new_document);
    }

    #[test]
    fn recognizes_frontend_menu_commands() {
        assert!(!is_frontend_menu_command("newDocument"));
        assert!(is_frontend_menu_command("openDocument"));
        assert!(is_frontend_menu_command("closeDocument"));
        assert!(!is_frontend_menu_command("openFolder"));
        assert!(is_frontend_menu_command("saveDocument"));
        assert!(is_frontend_menu_command("exportPdf"));
        assert!(is_frontend_menu_command("exportHtml"));
        assert!(is_frontend_menu_command("formatBold"));
        assert!(is_frontend_menu_command("insertImage"));
        assert!(is_frontend_menu_command("insertTable"));
        assert!(is_frontend_menu_command("toggleMarkdownFiles"));
        assert!(is_frontend_menu_command("toggleAiAgent"));
        assert!(is_frontend_menu_command("toggleAiCommand"));
        assert!(is_frontend_menu_command("toggleSourceMode"));
        assert!(!is_frontend_menu_command("markra:file"));
        assert!(!is_frontend_menu_command("copy"));
    }

    #[test]
    fn recognizes_native_new_window_menu_command() {
        assert!(is_native_new_window_command("newDocument"));
        assert!(!is_native_new_window_command("saveDocument"));
    }

    #[test]
    fn recognizes_native_settings_window_menu_command() {
        assert!(is_native_settings_window_command("openSettings"));
        assert!(!is_native_settings_window_command("saveDocument"));
        assert!(!is_frontend_menu_command("openSettings"));
    }
}
